scipy/special/specfun.c

```diff
@@ -0,0 +1,8418 @@
+        for (k = 1; k <= 12; k++) {
+            cr *= -0.25*(4.0*k + 1.0)*(4.0*k - 1.0)/zp2;
+            cg += cr;
+        }
+        cg *= 1.0/(pi*z*z);
+        s = d - (cf*ccos(zp) + cg*csin(zp))/(pi*z);
+    }
+    *zf = s;
+    *zd = csin(0.5*pi*z*z);
+    return;
+}
+
+
+double complex specfun_cchg(double a, double b, double complex z) {
+
+    // ===================================================
+    // Purpose: Compute confluent hypergeometric function
+    //          M(a,b,z) with real parameters a, b and a
+    //          complex argument z
+    // Input :  a --- Parameter
+    //          b --- Parameter
+    //          z --- Complex argument
+    // Output:  CHG --- M(a,b,z)
+    // Routine called: CGAMA for computing complex ln[Г(x)]
+    // ===================================================
+
+    int i, j, k, la, m, n, nl, ns;
+    double a0, a1, phi, x0, x, y;
+    double complex cfac, cg1, cg2, cg3, chg, chg1, chg2, chw, cr, cr1, cr2, cs1,\
+                   cs2, crg, cy0, cy1, z0;
+    const double pi = 3.141592653589793;
+    const double complex ci = CMPLX(0.0, 1.0);
+    a0 = a;
+    a1 = a;
+    z0 = z;
+    cy0 = 0.0;
+    cy1 = 0.0;
+    if ((b == 0.0) || (b == -(int)fabs(b))) { return 1e300; }
+    if ((a == 0.0) || (z == 0.0)) { return 1.0; }
+    if (a == -1.0) { return 1.0 - z/b; }
+    if (a == b) { return cexp(z); }
+    if (a - b == 1.0) { return (1.0 + z/b)*cexp(z); }
+    if ((a == 1.0) && (b == 2.0)) { return (cexp(z)-1.0) / z; }
+    if ((a == (int)a) && (a < 0.0)) {
+        m = (int)(-a);
+        cr = 1.0;
+        chg = 1.0;
+        for (k = 1; k < (m+1); k++) {
+            cr *= (a+k-1.0)/k/(b+k-1.0)*z;
+            chg += cr;
+        }
+    } else {
+        x0 = creal(z);
+        if (x0 < 0.0) {
+            a = b-a;
+            a0 = a;
+            z = -z;
+        }
+        nl = 0;
+        la = 0;
+        if (a >= 2.0) {
+            nl = 1;
+            la = (int)a;
+            a -= la + 1;
+        }
+        ns = 0;
+        for (n = 0; n < (nl+1); n++) {
+            if (a0 >= 2.0) { a += 1.0; }
+            if ((cabs(z) < 20.0+fabs(b)) || (a < 0.0)) {
+                chg = 1.0;
+                chw = 0.0;
+                crg = 1.0;
+                for (j = 1; j < 501; j++) {
+                    crg *= (a+j-1.0)/(j*(b+j-1.0))*z;
+                    if (cabs((chg-chw)/chg) < 1e-15) { break; }
+                    chw = chg;
+                }
+            } else {
+                y = 0.0;
+                cg1 = specfun_cgama(a, 0);
+                cg2 = specfun_cgama(b, 0);
+                cg3 = specfun_cgama(b-a, 0);
+                cs1 = 1.0;
+                cs2 = 1.0;
+                cr1 = 1.0;
+                cr2 = 1.0;
+                for (i = 0; i < 9; i++) {
+                    cr1 *= -(a+i-1.0)*(a-b+i)/(z*i);
+                    cr2 *= (b-a+i-1.0)*(i-a)/(z*i);
+                    cs1 += cr1;
+                    cs2 += cr2;
+                }
```

Comment:
```suggestion
                for (i = 1; i < 9; i++) {
                    cr1 *= -(a+i-1.0)*(a-b+i)/(z*i);
                    cr2 *= (b-a+i-1.0)*(i-a)/(z*i);
                    cs1 += cr1;
                    cs2 += cr2;
                }
```

Reply:
Indeed that's what I fixed but also before this one `CG2` becomes `inf` in the Fortran code already. So I couldn't understand how we get a finite result but anyways, that's a separate point I think we shouldn't waste time on.